core/util/src/main/java/org/eclipse/rdf4j/common/iteration/ConvertingIteration.java

```diff
@@ -56,10 +58,18 @@ protected abstract T convert(S sourceObject)
 	 * @return <tt>true</tt> if the source type iteration contains more elements, <tt>false</tt> otherwise.
 	 * @throws X
 	 */
+	@Override
 	public boolean hasNext()
 		throws X
 	{
-		return iter.hasNext();
+		if (isClosed()) {
+			return false;
+		}
+		boolean result = iter.hasNext();
+		if (!result) {
+			close();
```

Comment:
I may be wrong (been working on this for a while), but I think there are already self-closing cases in the code before this pull request. 

I also don't accept the general premise that we are hurting anything, especially users, in general by halting iterations either when hasNext will be returning false or just before next() would throw a NoSuchElementException. If a CloseableIteration knows that it can't immediately deliver a result, it should either block on hasNext or not respond to the close request as MutableTupleQueryResult does. If it doesn't want to follow the general contract that if hasNext ever returns false it is all over it is free to do so but we shouldn't encourage resource leaks just to cater for some situations where the programmer doesn't know their codebase well enough to know where to start looking when they accidentally exhause iterations and actually get a stack trace to help them start debugging it.

I don't accept the specific case that close on exhaustion can create deadlock bugs due only to the closure and would need specific examples to convince me of that one.

Relying on a complex finalizer mechanism instead of cleaning up after oneself at the time when the contract says we are free to do so is troublesome and will require more memory (while things leak onto the heap before getting finalised later) and time (both developer and running time) determining the source of bugs. Having said that, users are free to maintain a finalizer mechanism to pick up the odd case where this doesn't succeed, but it isn't a clean solution to resource management. Deadlocks are much more likely to happen in finalizers, due to having to get monitors on other threads, compared to when the cleanup is done within an objects own code where the immediate monitor is already guaranteed by the JVM to be held and any necessary locks are in place and don't need to be renegotiated.

I can imagine more cases where the source of the intermittent bug behind this pull request (that of HttpClient internals being modified outside of the handleClose method), that would be exacerbated by relying on lazy/queue-driven/thread-driven finalisation compared to active early closing.

FWIW, the only failures I have noticed with any of these changes are the Lucene failures that are not replicable and not unique to this so cannot be directly related to these changes. All of our other tests have been rock-solid throughout the development of this pull request, based on running full builds after ever 1-3 commits on my local machine given the complexity as a whole of the pull request.

If someone wanted to avoid the autoclose behaviour where they were using a subclass of AbstractCloseableIteration in their own code (say one of the query results iterators) but didn't want it to close just because it got to the end of its iteration, all they need to do is override the handleClose method to do nothing and the calls to close will silently short-circuit after going into handleClose the first time and then be avoided everytime after that. 

This pull request is just about being consistent internally and following the stated contract for CloseableIteration internally to avoid as many resource leaks and intermittent bugs in the future as possible.

I don't mind having a discussion about this, but fear of the unknown or anecdotal cases isn't a reason not to engineer it well IMO. The lack of consistency in the handleClose pattern (given that all of these iteration classes are used and stacked all over the API) triggered a few other changes attached to this, but they can be removed if someone can show they cause a bug in a minimal test case that can't be fixed by more cleanup.